packages/pricing/src/services/pricing-module.ts

```diff
@@ -79,46 +97,90 @@ export default class PricingModuleService<
     pricingContext: PricingContext = { context: {} },
     @MedusaContext() sharedContext: Context = {}
   ): Promise<PricingTypes.CalculatedPriceSetDTO> {
+    const manager = sharedContext.manager as EntityManager
+    const knex = manager.getKnex()
+
     // Keeping this whole logic raw in here for now as they will undergo
     // some changes, will abstract them out once we have a final version
     const context = pricingContext.context || {}
-    const priceSetFilters: PricingTypes.FilterablePriceSetProps = {
-      id: pricingFilters.id,
+
+    // Quantity is used to scope money amounts based on min_quantity and max_quantity.
+    // We should potentially think of reserved words in pricingContext that can't be used in rules
+    // or have a separate pricing options that accept things like quantity, price_list_id and other
+    // pricing module features
+    const quantity = context.quantity
+    delete context.quantity
+
+    // Gets all the price set money amounts where rules match for each of the contexts
+    // that the price set is configured for
+    const psmaSubQueryKnex = knex({
+      psma: "price_set_money_amount",
+    })
+      .select({
+        id: "psma.id",
+        price_set_id: "psma.price_set_id",
+        money_amount_id: "psma.money_amount_id",
+        number_rules: "psma.number_rules",
+      })
+      .join("price_rule as pr", "pr.price_set_money_amount_id", "psma.id")
+      .join("rule_type as rt", "rt.id", "pr.rule_type_id")
+      .orderBy("number_rules", "desc")
+
+    for (const [key, value] of Object.entries(context)) {
+      psmaSubQueryKnex.orWhere({
+        "rt.rule_attribute": key,
+        "pr.value": value,
+      })
     }
 
-    const priceSets = await this.list(
-      priceSetFilters,
-      {
-        select: [
-          "id",
-          "money_amounts.id",
-          "money_amounts.currency_code",
-          "money_amounts.amount",
-          "money_amounts.min_quantity",
-          "money_amounts.max_quantity",
-        ],
-        relations: ["money_amounts"],
-      },
-      sharedContext
-    )
+    psmaSubQueryKnex
+      .groupBy("psma.id")
+      .having(knex.raw("count(DISTINCT rt.rule_attribute) = psma.number_rules"))
+      .andHaving(knex.raw(`${Object.entries(context).length} > 0`))
```

Comment:
If there isn't any context, and we are not yet using a UNION ALL to get prices with no rules associated with them and no dynamic rules, I believe we can completely avoid to query the database.